### DISCUSSION TITLE
"StreetComplete for experienced mappers"

### DISCUSSION THREAD

**Discussion by smichel17** [streetcomplete/StreetComplete#3003] [Ideas]
It seems like there are two groups of mappers using SC in different ways:

- People who (almost) **only** contribute to OSM via StreetComplete. I'd put myself solidly in this category. Around 3k :star:s with SC puts me around the top 5% of SC users, I think; I can count my other OSM edits on my fingers. And I intend to keep it that way, partially due to lack of interest, partially to preserve my perspective as an outsider.

    - For me, the main attraction of SC is that it saves me from having to learn details of OSM tagging. It gives me something to do on walks, and some purpose. Like a more meaningful Pokemon Go.

- People who use other editors and are perhaps more involved in the OSM ecosystem. "Experienced mappers", for lack of a better term. I think the majority of regulars in the comments here fall into this category.

    - I can't really speak for you all, but I imagine that it's similar and different from me. Similar = also partially "something to do on walks". Different = an efficient tool for adding information that requires surveying.

Most of the time these use cases overlap remarkably well, but I've noticed some times recently where they diverge. For example, in https://github.com/streetcomplete/StreetComplete/issues/3002, it is a totally reasonable assumption that any note or comment _of mine_ was created via SC, but for someone using other editors, this is no longer true.

I'm wondering if it would make sense to make a "StreetComplete for experienced mappers". It wouldn't necessarily have *more* settings or features, but it would be built with different assumptions about its use. Not sure of the right way to implement it — could be a secret toggle like Android developer settings, a separate build flavor, or a soft fork, like [Zazolc](https://github.com/matkoniecz/Zazolc) (~~which would probably be a good starting point for a project like this~~ see [below](https://github.com/streetcomplete/StreetComplete/discussions/3003#discussioncomment-932097)).

Also not sure if there are enough differences to make this worth doing, but I figure there are enough to open a discussion and talk about it :)

**Comment by matkoniecz** [comment]
> or a soft fork, like Zazolc (which would probably be a good starting point for a project like this).

No it would not be :) There may be for example quests that will crash SC if you answer them (it is intended only to signal that I should edit with Vespucci) and cryptic text that will be understood only by myself and tags without broad community support and experimental not-yet-working quests that sometimes apply buggy edits.

This is intended as a personal fork and there is no support whatsoever for other using it (though feel free to look at code and may use it in your own fork if you understand the code!).

If someone wants their own fork - I recommend branching from SC itself.

**Reply by smichel17** [reply]
I *did* say "starting point", but point taken :)

**Reply by rugk** [reply]
> There may be for example quests that will crash SC if you answer them 

It's a feature, not a bug! Just add a button "Do not crash" as an optional alternative. :wink: 

![](https://media.giphy.com/media/LZQsVAzgB6sE0/giphy.gif?cid=ecf05e47x06fp7y2vpnjl8n2nw57hwq09lmipbhasbmhia1w&rid=giphy.gif&ct=g)

**Comment by kmpoppe** [comment]
Even though I've only started to really contribute to the map myself a few years ago, I would count myself into the experienced users as per @smichel17 's definition.

SC is a wonderful tool, because it removes almost all barriers to updating the map. I for one don't run around with my desktop PC on a trolley so an editor that does exactly what's required to correctly update an object is wonderful to have at hand.

Anyhow, I would be _in favour_ of an expert mode as in: Have a button that shows me all tags of the object in question in a table view and lets me change them. Why? let's say I get question if the opening hours of the shop "ABC Lettershop" are still correct. Now they might, yet the shop's now called "XYZ Appliances". Today, I'd have to leave a Note and change it later.

Going one step back into what's already on SC another way could be to list all quests that WOULD fit the object (like shop=*, amenity=*) and give the user a tick box to indicate which information need updating and then proceed as if the quest had not been answered. 

These are just very rough ideas and I'm well aware that @westnordost has very limited time available for these major changes but bouncing ideas can never hurt, can it?

Kai

**Reply by mnalis** [reply]
There already is such feature, just `long press` on location you're interested in and choose `open location in other app` and choose `Vespucci` or your other full-fledged editor of choice. Also, see https://github.com/streetcomplete/StreetComplete/issues/471 for discussion and resolution on such `Expert mode`.

> bouncing ideas can never hurt, can it?

Theoretically, If devels are subscribed to discussions on their projects, sure it can hurt, as there is only 24h in a day :smile: But unlikely to become an issue, so feel free to throw ideas around!

**Comment by smichel17** [comment]
I also wonder whether an updated UI for Vespucci might be a better route than a different flavor of StreetComplete. What is it about SC that keeps you folks using it (presumably often *alongside* Vespucci)?

**Reply by mnalis** [reply]
> I also wonder whether an updated UI for Vespucci might be a better route than a different flavor of StreetComplete. What is it about SC that keeps you folks using it (presumably often _alongside_ Vespucci)?

I don't think so, unless Vespucci reimplemented whole of SC into itself.

SC is being much **less cluttered visually**, there is **no chance of accidentally changing something unrelated**, it's having **questions pre-made**, and most importantly, **editing is much quicker** with no need to think. For example in StreetComplete:
- I'm presented with `lit` quest so I click on it
- I click on `yes`

This is it. 2 clicks and no thinking nor researching needed. In Vespucci:
- determine what needs to be edited (see [directed vs. non-directed](https://github.com/streetcomplete/StreetComplete/discussions/2975#discussioncomment-889913) discussion). You can activate Osmose in Vespucci to indicate things that are broken/missing, but it is not the same thing as SC does, so won't help in this specific example. **This can be a lot of work in itself** if you don't have particular idea of what exactly you want to edit but only a general vague feeling of _"wanting to make the map better"_.
- click `unlock` button
- (extra zoom in, as there are much more things that Vespucci can edit than StreetComplete)
- click on street
- (choose that you want to edit the street, and not some of relations it is in)
- click on `edit` (pen) button
- Notice it does not have `lit` tag which you want
- click on `+abc`
- (scroll and) select checkbox on the `Lit` tag to indicate it is lit
- confirm edit with `checkmark` icon
- click `back` icon to unselect way
- click `lock` icon again to prevent yourself  from breaking stuff while panning/zooming
- (If you have autoupload enabled in SC, there there is more clicks to actually upload data, but it is per-session and not per-quest, so no biggie)

The reason why is SC preferred for any of the stuff it supports should be clear even without reading and just visually comparing the paragraphs above. :smile: Also, you can use StreetComplete with gloves on in winter, something I would never recommend in Vespucci (it needs more precision).

On the other hand, when adding elements especially ways, polygons etc, or when needed full power of editing tags, Vespucci is undisputed mobile king. If you have mobile phone with you, you can practically do anything you can on a desktop.

**Reply by matkoniecz** [reply]
Excellent summary - see also opening hours editors in StreetComplete and Vespucci.

Vespucci one allows to enter bizarre and complex rules - but at cost of making harder to add simple cases.

**Comment by Helium314** [comment]
What should actually be in such an "expert edition"?

Some ideas
* edit tags directly
* per-quest settings
  * element selection (e.g. to stop asking for sidewalks on trunk roads, or ask for roof shape of higher buildings)
  * answers (e.g. choose which maxspeed tag to use)
* additional quests that don't have tags for all answers (e.g. telephone number for shop)
* show POIs
* create notes in gpx file
* quest filters / presets / profiles
* "invert quest order" button

(I have some of these in a more or less working state in my personal edition)

**Reply by smichel17** [reply]
> edit tags directly

Maybe instead, add "Edit [way|node] in other application" to the "Can't say" menu for quests. That way SC does not need a UI for editing tags directly, but it's easy to jump to Vespucci when something needs tweaking. I wonder if there is a way to open in Vespucci with the specific element pre-selected or if that's something which could be added.

**Reply by mnalis** [reply]
@Helium314 hey, it would interesting if you could document somewhere on https://github.com/Helium314/StreetComplete what changes from upstream you have working / mostly working, as others might be interested to try them out! (I know I am)

@smichel17 I do not think that [geo: URI scheme](https://en.wikipedia.org/wiki/Geo_URI_scheme) supports things like that, but Vespucci maybe could implement auto-selecting element if `geo:` coordinates look precise enough and there is OSM element with those exact coordinates (should ask for such feature there, though). Would probably work only for nodes, though (finding center of polygons and comparing seems much more daunting task)

**Reply by Helium314** [reply]
> @Helium314 hey, it would interesting if you could document somewhere on https://github.com/Helium314/StreetComplete what changes from upstream you have working / mostly working, as others might be interested to try them out! (I know I am)

What I have:
* POI display for some types (sometimes not displaying existing POIs)
* contact quests (website and telephone)
* disable auto-download if auto-upload is disabled (because sometimes auto-download is annoying and loads uninteresting areas)
* higher location check frequency (otherwise my phone sometimes delivers really bad precision)
* adjusted dark theme for higher contrast
* button to temporarily reverse quest order (find "hidden" quests without hiding or adjusting quest order)
* gpx notes (without images)
* profiles for quest visibility and order (somewhat buggy)
* some small modifications to quests (no sidewalks for trunk, more buildings for which levels are asked, allow paved surface without adding a note, different order for cycleways answers).

will put a version for testing on github in the next few days

**Reply by Helium314** [reply]
code is available: https://github.com/Helium314/StreetComplete/tree/mods
some more relevant information for usage:
* you can add images when creating gpx notes, but they will be ignored
* for nicer display of POI quests I moved the "quest dots" closer to the actual position of the quest, so the dot will not be hidden by the pin any more.
There is no information what is displayed by each of the POI quests except in the code: https://github.com/Helium314/StreetComplete/tree/mods/app/src/main/java/de/westnordost/streetcomplete/quests/show_poi
* reverse quest order button works (shown as plus/minus above _undo_ button), but quest order will be reset to normal if you enter the settings screen. This leaves the button in the incorrect state, but it still works
* profiles are buggy: after initial modifications it can/will happen that _all_ quests are disabled. It's necessary to reset quests to default before using the profile after first modification and leaving quest order screen (notes quest can't be re-enabled in any other way)

**Reply by smichel17** [reply]
@mnalis I think (?) it should be do-able without a custom protocol if Vespucci were extended to support OSM links, like https://www.openstreetmap.org/way/802075510#map=13/39.2446/-99.4989

edit: actually I guess https://www.openstreetmap.org/edit?way=802075510#map=13/39.2446/-99.4989 would be better

@westnordost Is a url in that form something which is feasible for SC to generate?

**Reply by mnalis** [reply]
@smichel17 custom protocol instead of `geo:` is probably always an option, but then, all the apps that want to use it must write support for it, or it won't work at all - so SC then should offer both `geo:` for non-specific apps, and another protocol (like currently only desktop based) [remote-control](https://wiki.openstreetmap.org/wiki/JOSM/RemoteControl) for apps-specifically-tailored-to-work-with-SC. Or one can work on extending `geo:` protocol... 

Anyway, does not look like something that can be solved by SC alone, but only if at least two (or even better, more!) interested parties see mutual benefit there.

**Reply by smichel17** [reply]
@mnalis I think you may have misunderstood me. I agree a custom protocol would not be as good, and I was proposing using standard OSM links instead. And yes, Vespucci would need to implement a handler for those, but this is probably a reasonable thing to do anyway. And if the user does not have Vespucci installed, it will nicely fallback to iD, which is admittedly not usable from mobile, but it's at least something reasonable (and you could share the url to a computer, or something).

**Reply by mnalis** [reply]
> code is available: https://github.com/Helium314/StreetComplete/tree/mods

@Helium314 do the prebuilt APKs work, though? I've been trying to install  `SC31_POI-quest.apk` and `sc_gpxnotes.apk` from https://github.com/Helium314/StreetComplete/releases, and I'm getting parse error `There was a problem parsing the package.` Perhaps the APK was [renamed after signing](https://stackoverflow.com/a/1501361/2600099) ?
Also, if you enabled `Issues` at your repository, people could give feedback there?

**Reply by Helium314** [reply]
Not sure, probably I didn't rename the package and it conflicts with an existing installation of SC.
I always rename my apk files after they are signed and never had a problem so far.

I see you have you own SC version at https://github.com/mnalis/StreetComplete/tree/mnalis, so you could simply add what you want from https://github.com/Helium314/StreetComplete/commits/mods. I took care to have the commits independent from each other.

But be aware that you'll probably have to reset the quest order/visibility if you include the profiles.

**Reply by Helium314** [reply]
Had some time, apk is available now at https://github.com/Helium314/StreetComplete/releases/tag/v33.1
The package name is different from SC, so it should be possible to install without removing the original version.

> Also, if you enabled `Issues` at your repository, people could give feedback there?

TBH I don't intend to change/improve what I have, at least not in the near future. This is because it's going to be rather time consuming and iend up getting used by maybe 2-3 people.
So feedback is welcome, but it's unlikely I will change anything based on the feedback.

However, if this "experienced mapper edition" becomes a real thing, I'd definitely try to include some of my mods, and adjust them according to feedback.

**Reply by mnalis** [reply]
@Helium314 thanks, new prebuilt `.apk` installs fine now! I'm still having app crashes during initial resetting of the profiles, but once setup they seem to work (so far). I'll play some more to see how well it handles.

**Reply by andrewharvey** [reply]
@Helium314 are you accepting PRs? I'd like to add some quests which SC mainline either rejected or I feel probably wouldn't be accepted. `addr:flats`, `building:flats`, loosen quest criteria so less quests are filtered out, sidewalk/footway/cycleway width, gate access, locked gate, track access, indoor/outdoor seating, tree species, traffic signals ref, power pole ref, parking maxstay,

**Reply by Helium314** [reply]
@andrewharvey 
currently I intend to have my version "read-only", main reason is that I definitely don't feel ready to maintain that proposed "experienced mapper edition" where such changes should be added.

**Reply by Helium314** [reply]
I added 2 more things that might be useful, see https://github.com/Helium314/StreetComplete/commits/mods
* hide button for faster hiding
This should be especially convenient for all the quests that don't have an answer
Currently blocks a part of the "other answers" button, but it's still possible to tap.
* per-quest settings
I didn't receive any feedback the 2 or 3 times I suggested it, but still decided to implement it as it can be used to change element selection. Currently this is possible for roof shape (show for high buildings), building level (select building types) and place name (add arbitrary elements, e.g. for #925). It should be easy to extend it, e.g. tfor adding address to bungalows, or change the maxspeed tag that gets added.
However, this feature is a bit dangerous: a) it's probably possible to crash the app via bad selection strings, and b) it's _much_ easier now to do incorrect tagging

**Reply by smichel17** [reply]
> What should actually be in such an "expert edition"?

Perhaps a batch mode? e.g. https://github.com/streetcomplete/StreetComplete/issues/3098

**Reply by Helium314** [reply]
Another update, adding optional zoom with volume buttons and a quest asking for level of some POIs inside malls and retail buildings.

I have the impression there isn't much interest in such a version, so I don't really want to keep updating this version...

What I could do is simply publish the exact version I am using, which includes some additional small adjustments tuned for my personal taste.
Then I would be open for issues / PRs regarding experimental features or quests that are not suitable for SC.

**Reply by mnalis** [reply]
@Helium314 could you add `README-Helium314.md` to your repository listing changes/features that you've added? It is getting hard to keep track. Then it would likely be just fine to publish whatever you are using. Just keep different improvements in different commits... 

Interested parties can then download `.apk` and see how it works and suggest what could be improved. People who want to test some experimental/expert feature usually won't mind that some background color is also changed (or whatever other small adjustments you have made for your personal taste) etc. 

That's what I do, download your (or some other) version and see/test what I would like, and then `git cherry-pick` those into my fork (I still have separate fork, as for some of the features what I want seems to be in minority, like for example https://github.com/streetcomplete/StreetComplete/issues/2692#issuecomment-856073050)

If some experimental feature become polished enough that it might be considered for inclusion in upstream SC, then new branch/PR can be created for just that from existing commits.

**Reply by Helium314** [reply]
Issues activated, readme updated: https://github.com/Helium314/StreetComplete/tree/my_modifications

**Reply by dlichtistw** [reply]
> What should actually be in such an "expert edition"?

- User Defined Quests: The user can enter a filter string and a list of add/update/remove tag commands to define custom quests. Complexity would be limited, but it could help in prototyping new quests and sharing them between "experts".

**Comment by westnordost** [comment]
High contrast dark mode sounds interesting. Got pics?

On July 3, 2021 12:08:22 PM GMT+02:00, Helium314 ***@***.***> wrote:
>> @Helium314 hey, it would interesting if you could document somewhere
>on https://github.com/Helium314/StreetComplete what changes from
>upstream you have working / mostly working, as others might be
>interested to try them out! (I know I am)
>
>What I have:
>* POI display for some types (sometimes not displaying existing POIs)
>* contact quests (website and telephone)
>* disable auto-download if auto-upload is disabled (because sometimes
>auto-download is annoying and loads uninteresting areas)
>* higher location check frequency (otherwise my phone sometimes
>delivers really bad precision)
>* adjusted dark theme for higher contrast
>* button to temporarily reverse quest order (find "hidden" quests
>without hiding or adjusting quest order)
>* gpx notes (without images)
>* profiles for quest visibility and order (somewhat buggy)
>* some small modifications to quests (no sidewalks for trunk, more
>buildings for which levels are asked, allow paved surface without
>adding a note, different order for cycleways answers).
>
>will put a version for testing on github in the next few days
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/streetcomplete/StreetComplete/discussions/3003#discussioncomment-959591

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Reply by Helium314** [reply]
![modified](https://user-images.githubusercontent.com/43007630/124354319-bd15b700-dc0b-11eb-9b06-78da7a3bd04e.png)
![original](https://user-images.githubusercontent.com/43007630/124354321-be46e400-dc0b-11eb-91a5-97053e2a3d16.png)

two main reasons for this:
* more contrast between "generic" ground (now darker) and buildings (edges now much brighter), which helps a lot when viewing my screen in bright sunlight
* more "natural" colors for water, grass and forest areas to make them easier recognizable

**Reply by smichel17** [reply]
I like the changes on land. In particular, the area in the middle of the top center square building looks blue, like water, in the lower shot; in the upper shot it is clearly land. The increased contrast of buildings also makes the small buildings (which I presume are sheds) just below easier to spot.

I'm lukewarm on the water change. I think it would be good to make it a little more blue, but I think your change went too far. In particular, I don't like the area where the bridge crosses. Maybe somewhere in between the two?

**Reply by Helium314** [reply]
Note that the colors are adjusted for my phone screen, not any other.
On my computer screen it definitely looks different, probably similar to what you see.
But on my phone I get a nice dark blue.

**Comment by smichel17** [comment]
Other issues like https://github.com/streetcomplete/StreetComplete/issues/3084#issuecomment-883693413 (cc @matkoniecz @Helium314) have me more convinced that this would be a good idea. It's silly for multiple people to independently maintain the same feature in their own private forks.

I think the defining feature of this version of SC is that it's okay to wade into the "unclear" space. For example, the SC criteria "🤔 Any answer the user can give must have an equivalent tagging (Quest should not reappear to other users when solved by one)" does not need to apply, because the user is expected to know when there is no clear answer or the data needs to be fixed from a different editor.

I think this version would be useful as a playground/prototype/test-bed for SC, where there is not so much focus on polish or code quality. Then if a feature from this version seems useful more generally, it can be polished up to SC standards.

**Reply by Helium314** [reply]
So we have some ideas of what would be in this version, and we already have some of the code.

We need some github repo to put the code, someone willing to keep it up to date with SC, and some people who actually want to use it (didn't get much feedback on my apks so far, no idea wheter anyone except mnalis actually tried it)

**Comment by Helium314** [comment]
In case anyone is interested:
New release available at https://github.com/Helium314/StreetComplete/releases/tag/v34.0b1

Modifications inspired by issues/discussions:
* added quests
  * ask for service building operator (#2859)
  * ask for level of places inside buildings (#1487)
  * ask for phone and website of certain places (#2043, #2839)
  * ask for cuisine of restaurants (#103, #1938)
* quest changes
  * add "private" other answer to some road/path quests (#2051)
  * add "demolished" other answer to building quest (#2719)
  * move "none" answer for cycleway quest to more accessible position (sort of related to #2651)
  * add "zebra" answer to crossing quest (#2087)
  * per-quest settings for some quests, e.g. allows choice of used maxspeed tag
* show certain POIs and fixmes as large clickable dots (#844, #3084)
* zoom using volume buttons (optional, #3095)
* show quest geometries (optional, #1576)
* disable the new daytime filter (optional)
* quick settings menu
  * switch profiles
  * reverse quest order (easier access to quests "behind" other quests)
  * level filter (#2874)
* show "hide" button like "ok" button (#3023)
* allow adding notes to gpx file (without photos, #1992)
* use last known location as initial location (if less than 30 seconds old and reasonably precise)

For list of more complete list of changes see readme at https://github.com/Helium314/StreetComplete

**Comment by mnalis** [comment]
It was mentioned before in some issue, but perhaps such "expert mode"  should be unlocked only after user having proven not to be a newbie - maybe by looking username up at [HDYC](https://hdyc.neis-one.org/) ?

**Comment by jidanni** [comment]
All I know is if the boss sends me out on a mission to check and correct
all electric pole numbers on rural road 33A, and during so I notice that
there is one pole entirely missing from the map, it would be great to
have the ability to add things, not only correct things.

**Reply by mnalis** [reply]
You can always long press at that location and leave a note (with picture) and then later enter it into the map. Or you can long press and open another more full-blown editor (like [Vespucci](https://wiki.openstreetmap.org/wiki/Vespucci)) at that location.
 
Problem with adding things to the map is that you would need to know were all the other similar things are to be sure you're not adding duplicates. You would also need not only to add power tower, but also connect it to the correct power line, put it into [correct relation](https://wiki.openstreetmap.org/wiki/Washington#Power_Network) etc. Perhaps something like that might become possible with [Layers](https://github.com/streetcomplete/StreetComplete/issues/2461) idea. 
I still think Vespucci is your better bet for such complex tasks.

**Reply by jidanni** [reply]
I was just thinking about simple wooden power poles. I could just quickly copy a nearby one, and then change the plate number, and then enter it onto the map. They don't have a line connecting them to worry about, at least on the map.

**Reply by mnalis** [reply]
@jidanni Oh. People over here who care about the power poles care as much (or more) about the power lines connecting them. Out of curiosity, does your power company (?) depends on OSM for its database of power poles, or do you have your own database overlayed on top of OSM basemap (like GPX points inside OsmAnd or some other solution)?

**Reply by jidanni** [reply]
The power company makes pubic a list of poles and their coordinates.
https://lists.openstreetmap.org/pipermail/talk-tw/2021-October/001478.html
Wires connecting them are not public.
The power company isn't very familiar about OSM.

**Reply by mnalis** [reply]
If they publish under ODbL, perhaps the data could be [imported](https://wiki.openstreetmap.org/wiki/Import), probably using some [Conflator](https://wiki.openstreetmap.org/wiki/Conflation) to catch already mapped nodes? And thus avoid all that manual work of mapping them on the ground.

For entering random POIs while mapping on the ground, I usually use [OsmAnd](https://wiki.openstreetmap.org/wiki/OsmAnd) or [Vespucci](https://wiki.openstreetmap.org/wiki/Vespucci) (as StreetComplete can only leave notes, which requires manual postprocessing later, which is more work)

**Comment by Helium314** [comment]
Some summary after having my version on github for ~6 months, with issues and PRs enabled as requested in this discussion:
* one issue about a bug with satellite background (which I ended up disabling after it got worse in later versions)
* no PRs (expected from prior comments: quests not acceptable for StreetComplete and maybe experimental "enhancements")
* no further feedback (but at least some people were testing it)

Given the low amount of feedback / reactions, I'm actually glad I decided to simply publish my own version, containing only very few and simple things I don't use regularly, instead of attempting to create an "experienced mapper edition" with more features.

**Reply by mnalis** [reply]
Just to note (for people stumbling on this discussion) that since that time, more interest in advanced features have been shown, and it did result in @Helium314 making it more generally targeted (i.e. not just for personal use) [SCEE](https://wiki.openstreetmap.org/wiki/SCEE) (_"StreetComplete Expert Edition"_) fork.
